timelock-server/src/main/java/com/palantir/atlasdb/timelock/lock/LockAcquirer.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.lock;
+
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LockAcquirer {
+
+    private static final Logger log = LoggerFactory.getLogger(LockAcquirer.class);
+
+    public CompletableFuture<HeldLocks> acquireLocks(UUID requestId, List<AsyncLock> locks) {
+        CompletableFuture<Void> future = acquireLocksInOrder(requestId, locks, false);
+        registerErrorHandler(future, requestId, locks);
+
+        return future.thenApply(ignored -> new HeldLocks(locks, requestId));
+    }
+
+    public CompletableFuture<Void> waitForLocks(UUID requestId, List<AsyncLock> locks) {
+        return acquireLocksInOrder(requestId, locks, true);
+    }
+
+    private CompletableFuture<Void> acquireLocksInOrder(UUID requestId, List<AsyncLock> locks,
+            boolean releaseImmediately) {
+        CompletableFuture<Void> future = CompletableFuture.completedFuture(null);
+        for (AsyncLock lock : locks) {
+            future = future.thenCompose(ignored -> lock.lock(requestId));
+            if (releaseImmediately) {
+                future.thenRun(() -> lock.unlock(requestId));
+            }
+        }
```

Comment:
I checked this against the current `LockServiceImpl`. If I'm not wrong the current behaviour is as follows:

* If a lock is already available at the time we're trying to acquire it, then it just passes (`LockServiceImpl:453-456`)
* If it is not we lock it first _and then_ unlock all of them at the end (`LockServiceImpl:429-431`)

For illustration, suppose we have 3 threads T1, T2 and T3.

```
T2.lock([B]) locks B
T3.lock([D]) locks D
T1.lock([A, B, C, D, E]) - block indefinitely and release.
  A is acquired and then released.
  We block on B.
T2.unlock([B])
  B is acquired, but *NOT* released.
  C is acquired and then released.
  We block on D.
T3.unlock([D])
  D is acquired, but *NOT* released.
  E is acquired and then released.
  B and D are released.
```

So the existing service doesn't guarantee that we were holding all of the locks at any single point in time. On that basis I think this could be safe, though not entirely sure - @carrino for SA / thoughts.